### ISSUE TITLE
Palette node search stops working in desktop Safari

### ISSUE THREAD

**Issue by rcarmo** [node-red/node-red#4926]
### Current Behavior

After adding a node to the canvas the palette search/filtering stops working and I get a completely blank palette (where search does nothing) until I refresh the entire front-end on the browser
<img width="434" alt="image" src="https://github.com/user-attachments/assets/07d9840b-94b7-44cb-a38a-ff9a41395e63">

There are no errors in the JavaScript console.

### Expected Behavior

I expected search to keep working

### Steps To Reproduce

Tested on Safari Version 18.0.1 (20619.1.26.31.7) and node-red v4.0.5

### Example flow

```
paste your flow here
```

### Environment

- Node-RED version: v4.0.5
- Node.js version: 22.x
- npm version: N/A
- Platform/OS: Mac OS Sequoia 15.0.1 (24A348)
- Browser: Safari Version 18.0.1 (20619.1.26.31.7)

**Comment by hardillb** [comment]
Any errors in the browser dev tools console when it happens?

**Comment by rcarmo** [comment]
No, I specifically mentioned that. No adblocking or scripting extensions active for this side either.

**Comment by dceejay** [comment]
Confirmed. For me the search works ok - but when I try to undo it (clear the search) - it doesn't go back to the full list, and as per above the only way to reset is to reload the page.

**Comment by jsb5151** [comment]
Same issue here, works fine on Chrome Mac, but Safari is problematic.

**Comment by Cocoader** [comment]
I can also confirm it on Safari 18.0.1 (18619.1.26.111.11) / macOS 13.7.1 (22H221).

**Comment by p2cardon** [comment]
Same problem for me, MacOS 15.1.1, Safari 18.1.1, Nodered 4.0.3 under Home Assistant 2024.11
All works fine under MacOS Firefox
I thing the problem occurred since MacOS 15

**Comment by rcarmo** [comment]
This is still a thing in 4.0.8, released today.

**Comment by GogoVega** [comment]
Yes, we are aware of this, but the problem is that it's very difficult to isolate the source.

**Comment by bakman2** [comment]
> Yes, we are aware of this, but the problem is that it's very difficult to isolate the source.

I may not be knowledgeable enough, but every action taken in the search field triggers a webkit error in Safari for the clipboard:

<img width="954" alt="image" src="https://github.com/user-attachments/assets/caff8662-6e71-4fb3-a965-f13b99683b7d" />

**Comment by rcarmo** [comment]
@bakman2 that's because `navigator.clipboard` is, for security reasons, only available in specific contexts in Safari: https://webkit.org/blog/10247/new-webkit-features-in-safari-13-1/ (this is the only easy reference I could find). 

But I think that you've found a likely cause - that would seem to stop the rest of the search handling.

**Comment by dceejay** [comment]
That could make sense - but that code is in the Monaco editor handling library from Microsoft... :-(
( @Steve-Mcl ?)

**Comment by knolleary** [comment]
Palette search is nothing to do with Monaco.

**Comment by bakman2** [comment]
Safari has a requirement that the page is served over https in order to gain access the clipboard. Then again, i am accessing NR over https and the error still occurs.

**Comment by rcarmo** [comment]
Can confirm. I have multiple NR 4.x instances I routinely access, and this happens in all of them, HTTPS or not.

**Comment by bakman2** [comment]
Is there any possible solution/workaround for this issue ? Perhaps as a temporary fallback use the `document.execCommand('copy');` method for safari ? This method works.

**Comment by knolleary** [comment]
This needs someone who can reproduce this issue to debug properly. I don't currently have this version of Safari available to do that myself.

The errors shown above related to the clipboard are not coming from Node-RED's own code. Nor does the sidebar search/filter component do anything with the clipboard.

It does however touch localStorage since we added the ability to remember the filter/state of the palette.

Could someone who is able to reproduce this error, please run the following command in the javascript console and report back the result:

```
RED.settings.getLocal("palette-state")
```

What is odd is that all of our access to localStorage is already wrapped in try/catch blocks and would have logged an error.

**Comment by bakman2** [comment]
> Nor does the sidebar search/filter component do anything with the clipboard.

The error gets triggered by `/vendor/monaco/dist/editor.js` - i think that is what @dceejay was trying to point out. 

```
 installWebKitWriteTextWorkaround()
                    {
                        const e = () => {
                            const e = new Vt.CR;
                            this.webKitPendingClipboardWritePromise && !this.webKitPendingClipboardWritePromise.isSettled && this.webKitPendingClipboardWritePromise.cancel(),
                            this.webKitPendingClipboardWritePromise = e,
                            navigator.clipboard.write([new ClipboardItem({
                                "text/plain": e.p
                            })]).catch((async t => {
                                t instanceof Error && "NotAllowedError" === t.name && e.isRejected || this.logService.error(t)
                            }))
                        };
                        this._register(we.ju.runAndSubscribe(this.layoutService.onDidAddContainer, (({container: t, disposables: i}) => {
                            i.add((0, ve.nm)(t, "click", e)),
                            i.add((0, ve.nm)(t, "keydown", e))
                        }), {
                            container: this.layoutService.mainContainer,
                            disposables: this._store
                        }))
                    }
```

Interestingly enough, their fallback function below it actually uses the `exec` command as I suggested above, but I guess we don't have control over their package.

>

```
RED.settings.getLocal("palette-state")
{\"filter\":\"\",\"collapsed\":[]}
```
It does record anything I type in the palette search bar: 'inj'
```
RED.settings.getLocal("palette-state")
"{\"filter\":\"inj\",\"collapsed\":[]}"
```

**Comment by psy0rz** [comment]
so this will never be fixed??

**Comment by Steve-Mcl** [comment]
There is a Monaco PR waiting to be merged that may or may not fix the issue (assuming for a second the issue is related to the console error) but it still needs someone to pull that branch and test it.

**Comment by psy0rz** [comment]
update: it seems to have to do with Safari adguard. 

When i disable this extension, the menu works fine. If i enable it, it stops working after 1 or 2 refreshes.

Disabling adguard for the specific nodered page DOES NOT HELP. You have to disable the whole extension it seems.

So maybe nodered is not to blame, but perhaps it is fixable with this information?

**Comment by rcarmo** [comment]
Not really about ad blocking - at least in my case.

**Comment by anttiryt** [comment]
screen recording of the issue. 
Interesting to see how it affects both legacy node-red and v4 nodered, and how it appeared some time ago to Safari users.

https://github.com/user-attachments/assets/af37db61-18da-406e-8944-878f24327fb0

**Comment by rguca** [comment]
Just ran into this

**Comment by knolleary** [comment]
I can now (finally) reproduce this locally. Am debugging it.

**Comment by knolleary** [comment]
So, I can see why it's failing. When filtering nodes, we set the css `display: none` on the nodes that should be hidden. We then set `display: block` if they should be shown. After applying the filter, we can check the value of `display` of all nodes in each category to decide if the category should be shown or not - if all the nodes are `display: none` we keep the category hidden.

In all other browsers, after setting `display: block` on a node element inside a *hidden* category, the node element reports its `display` value as `block`.

In Safari, it reports` none` - despite the css being explicitly set to `block` when we examine the dom directly. This means it keeps the categories hidden.

Well done Safari.

We will need a different method to track visibility state that doesn't rely on CSS properties.

**Comment by knolleary** [comment]
Fixed by https://github.com/node-red/node-red/pull/5178
